DependencyInjection/Compiler/FirewallCompilerPass.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Scheb\TwoFactorBundle\DependencyInjection\Compiler;
+
+use Ivory\CKEditorBundle\Exception\DependencyInjectionException;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\DefinitionDecorator;
+use Symfony\Component\DependencyInjection\Reference;
+
+class FirewallCompilerPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container)
+    {
+        $authManagerDefinition = $container->getDefinition('security.authentication.manager');
+        $authProviders = $authManagerDefinition->getArgument(0);
+        $firewallServiceIds = $this->getServicesToDecorate($authProviders);
+
+        foreach ($firewallServiceIds as $firewall => $serviceIds) {
+            foreach ($serviceIds as $serviceId) {
+                $decoratedServiceId = $serviceId . '.two_factor_decorator';
+                $container
+                    ->setDefinition($decoratedServiceId, new DefinitionDecorator('scheb_two_factor.security.authentication.provider.decorator'))
+                    ->setDecoratedService($serviceId)
+                    ->replaceArgument(0, new Reference($decoratedServiceId . '.inner'))
+                    ->replaceArgument(1, $firewall);
+            }
+        }
+    }
+
+    private function getServicesToDecorate(array $authProviders)
+    {
+        $firewallServicesToDecorate = array();
+        $authProvidersPerFirewall = $this->getAuthProvidersPerFirewall($authProviders);
+        foreach ($authProvidersPerFirewall as $firewall => $authProviders) {
+            if (isset($authProviders['two_factor'])) {
+                unset($authProviders['two_factor'], $authProviders['anonymous']); // No need to decorate those authentication providers
+                if (!isset($firewallServicesToDecorate[$firewall])) {
+                    $firewallServicesToDecorate[$firewall] = array();
+                }
+                $firewallServicesToDecorate[$firewall] = array_merge($firewallServicesToDecorate[$firewall], array_values($authProviders));
+            }
+        }
+
+        return $firewallServicesToDecorate;
+    }
+
+    private function getAuthProvidersPerFirewall(array $authProviders)
+    {
+        $firewallProviders = array();
+
+        /** @var Reference $authProvider */
+        foreach ($authProviders as $authProvider) {
+            $authProviderId = (string) $authProvider;
+            if (!preg_match('#^security\.authentication\.provider\.([^.]+)\.([^.]+)$#', $authProviderId, $match)) {
```

Comment:
This is unlikely to change, but I share your concern, it's fragile.
Perhaps it's good enough for now, but if you want to make it more robust, you could give a try to the `security.firewall.map` service which takes all FirewallContext (per firewall object) references as arguments, themselves containing a `FirewallConfig` object and all configured listeners. Note that it's not the same before 3.3 so again, I suggest you to drop compatibility for sf <3.3, it's going to be complex and keeping compatibility would be a true nightmare. 
AFAIK, this code is already incompatible with 3.4+, e.g. because authentication providers are not injected as an array in the authentication manager, but wrapped in an `IteratorArgument` for laziness purpose, so you first need to get the array from the iterator via `$argument->getValues()`.

Reply:
I need to look into `FirewallConfig`, haven't worked with that yet. Migration to `IteratorArgument` is definitely something I'd do when I upgrade the branch to a current version.